### ISSUE TITLE
module not found: Can't resolve 'date-fns/esm/format'

### ISSUE THREAD

**Issue by sushilbansal** [marnusw/date-fns-tz#133]
1) npm i date-fns date-fns-tz
2) import { utcToZonedTime } from 'date-fns-tz';
utcToZonedTime(new Date(), timezone) 

Run the app and got this issue:
error - ./node_modules/date-fns-tz/esm/format/index.js:1:0
Module not found: Can't resolve 'date-fns/esm/format'

**Comment by nraj-willow** [comment]
I am also getting same issue

**Comment by ayhid** [comment]
same here :/

**Comment by marnusw** [comment]
Can you `import fromat from 'date-fns-tz/esm/format'` in your own code? I've just updated the dev dependency locally and `node_modules/date-fns-tz/esm/format/index.js` is definitely there. The tests also seem to work fine locally.

**Comment by davidcalhoun** [comment]
Something funky and similar happening in Node 14.17.5

```
mport { utcToZonedTime, format } from 'date-fns-tz'
         ^^^^^^^^^^^^^^
SyntaxError: Named export 'utcToZonedTime' not found. The requested module 'date-fns-tz' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'date-fns-tz';
const { utcToZonedTime, format } = pkg;

    at ModuleJob._instantiate (internal/modules/esm/module_job.js:121:21)
    at async ModuleJob.run (internal/modules/esm/module_job.js:166:5)
    at async Loader.import (internal/modules/esm/loader.js:178:24)
    at async Object.loadESM (internal/process/esm_loader.js:68:5)
```

**Comment by op** [comment]
When using `import pkg from 'date-fns-tz`, the `eslint-plugin-import` reports this error:
> import/default: No default export found in imported module "date-fns-tz"

I had to add `/* eslint-disable import/default */` to the mix to silence that error.

**Comment by alexwendte** [comment]
In my case I didn't have "date-fns" installed and it is a peer dependency of "date-fns-tz". Once I installed the "date-fns" I no longer had this issue and could just do: `import { utcToZonedTime } from 'date-fns-tz';` without a problem.

**Comment by marnusw** [comment]
It sounds like the solution to the OP is to install `date-fns`. I have added a note to the docs to make it clear that it is a peer depedency.

I am aware that the imports might not work as expected in modern ESM projects with `"type": "module"`. I will investigate this in due course, and anyone is welcome to open a separate issue to track this.

In the meantime you can try importing `date-fns-tz/esm/index.js`, although I'm not at all sure whether that will work. Most likely for bundlers like Webpack with Babel or TypeScript, but not directly in Node.js.

**Comment by bbo2adwuff** [comment]
Well, I just tried using `import { utcToZonedTime } from 'date-fns-tz';` in a project with `"type": "module"` with Node.js v17.3.0.

I tried the different suggestions from this thread here, the only suggestion that seems to work for me is the one from @davidcalhoun:

```js
import dateFnsTz from 'date-fns-tz';
const { utcToZonedTime } = dateFnsTz;
```

**Comment by marnusw** [comment]
Yes, the project doesn't work properly in type module projects. I'm pretty sure this can be solved using the `exports` property in package.json. I'll fix that when I have time. If someone for whom this is an issue will contribute that in the meantime it will be greatly appreciated.

**Comment by omgpiu** [comment]
ERROR in ./node_modules/date-fns-tz/esm/zonedTimeToUtc/index.js
Module not found: Error: Can't resolve 'date-fns/esm/_lib/cloneObject'

ERROR in ./node_modules/date-fns-tz/esm/zonedTimeToUtc/index.js 
Module not found: Error: Can't resolve '../toDate'

ERROR in ./node_modules/date-fns-tz/esm/zonedTimeToUtc/index.js 
Module not found: Error: Can't resolve '../_lib/tzPattern' 

ERROR in ./node_modules/date-fns-tz/esm/zonedTimeToUtc/index.js 
Module not found: Error: Can't resolve '../_lib/tzParseTimezone' 

It tries to find int  /node_modules/date-fns-tz/esm/zonedTimeToUtc.
I install date-fns, hacks with imports does not work for me.

**Comment by bbo2adwuff** [comment]
I also played around with what  @marnusw suggested (i.e. `exports` for `"type": "module"` projects).
And I had similar error messages to @omgpiu.

But shouldn't replacing `'../toDate'` with `'../toDate/index.js'`in following line work?
https://github.com/marnusw/date-fns-tz/blob/6b8ef0a5db8d893d8a36244881f64c80a4059460/src/zonedTimeToUtc/index.js#L2

Haven't time to try it out though...

**Comment by marnusw** [comment]
Yes, you will also need to update all imports to include file extensions.

**Comment by marnusw** [comment]
Please see whether `1.3.0-beta.0` solves this for you.

**Comment by bbo2adwuff** [comment]
As what node calls `main` was missing, adding `".": "./esm/index.js"` to `exports` helps.

Then I get the expected output, with any errors :rocket:
```
2022-02-11T08:00:00.000Z
```

Thanks for this great work :star_struck:

**Comment by marnusw** [comment]
Excellent! Thanks for investigating @bbo2adwuff. I've added that to `1.3.0-beta.1`. I will get a proper release out in the near future once I've done some clean-up.

**Comment by marnusw** [comment]
@bbo2adwuff that last change was bugging me, and it turns out for good reason. If the root import points to `./esm/index.js` the library doesn't work with CommonJS imports in the latest Node versions. Arguably many users will opt for ESM now that it's natively supported, but there will also be many legacy projects who are not ready to go with `"type": "module"`.

I have fixed both CommonJS and ESM as far as I can tell in `1.3.0-beta.3`. To use the library in an ESM project you will **have to** `import { zonedTimeToUtc } from 'date-fns-tz/esm'`. 

Tehnically this has always been the case, only importing from `date-fns-tz` directly worked because the bundlers would also pull in CommonJS modules. At least now Node.js will point out that the wrong module system is being used.

In a major release this could be swapped around so ESM becomes the default. It looks like `date-fns` might opt for something similar in v3.

I'll document this properly before the `1.3.0` realese. In the meantime, please let me know if you encounter any problems with this approach.

**Comment by bbo2adwuff** [comment]
True! Indeed I just checked how it works as ESM.
Now after your comment I tested both, CommonJS and ESM, and both work perfectly fine with `1.3.0-beta.3`.

And indeed many legacy projects use CommonJS. So do some of my projects, but for one I had to update to got `v12.0.0` which is [now pure ESM](https://github.com/sindresorhus/got/releases/tag/v12.0.0) and I got `Must use import to load ES Module` errors. So this is why I had to use ESM and commented on this issue :wink:.

Sounds like a good idea to also make ESM the default when `date-fns` does so :+1:

Thanks again for this great package! :rocket:
